feat(ui): restore has('gui_running')

justinmk opened the pull request:
In the past this was rejected because there is not a single GUI. But in practice, `has('gui_running')` is also used as a way to detect "is the TUI (not) running?" 

re https://github.com/vimpostor/vim-tpipeline/commit/95a6ccbe9f33bc42dd4cee45731d8bc3fbcd92d1

todo

- > Another solution could be to have the TUI set its name, so that plugins could check nvim_get_chan_info(1).client.name



bsd CI:

```
test/functional/terminal/tui_spec.lua:1427: Expected objects to be the same.
Passed in:
   ...
   *[term_name] = 'builtin_xterm'
```

justinmk committed:
feat(api): more fields in nvim_list_uis

Problem:
nvim_list_uis does not report all ":help ui-option" fields.

Solution:
Store ":help ui-option" fields on the `UI` object and update ui_array.

bfredl commented on src/nvim/eval/funcs.c:
```diff
@@ -3161,7 +3161,13 @@ static void f_has(typval_T *argvars, typval_T *rettv, EvalFuncData fptr)
   if (!n) {
-    if (STRNICMP(name, "patch", 5) == 0) {
+    if (STRNICMP(name, "gui_running", 11) == 0) {
+      char *term_name;
+      if (!get_tty_option("term", &term_name)) {
+        abort();
+      }
```
This doesn't do anything and can be removed

przepompownia commented:
I hope it's self-explaining assuming that I used binary built from this PR and ran it in tmux + kitty:

https://user-images.githubusercontent.com/11404453/221375768-a4ead278-a925-4736-890f-52833df75dbe.mp4

In short: manual command output shows the correct result in opposite to the output on `UIEnter`. 

With `nvim-qt --nvim ~/tmp/nvim-from-src/bin/nvim -- -u /tmp/init.lua` I get `1` both from autocommand and manual command.

justinmk committed:
feat(ui): restore has('gui_running')

Problem:
has('gui_running') is still common in the wild and our answer has
changed over time, causing frustration.
https://github.com/vimpostor/vim-tpipeline/commit/95a6ccbe9f33bc42dd4cee45731d8bc3fbcd92d1

Solution:
Use stdin_tty/stdout_tty to decide if a UI is (not) a GUI.

justinmk commented:
@przepompownia this PR was in draft so wasn't ready for testing. Now it is ready for testing :)

justinmk commented on src/nvim/api/ui.c:
```diff
@@ -351,6 +348,7 @@ static void ui_set_option(UI *ui, bool init, String name, Object value, Error *e
     });
 
     stdin_fd = (int)value.data.integer;
+    ui->stdin_fd = (int)value.data.integer;
```
not sure if this is useful in `nvim_list_uis` ?

justinmk commented on src/nvim/ui.c:
```diff
@@ -147,6 +148,18 @@ bool ui_rgb_attached(void)
+bool ui_gui_attached(void)
+{
+  for (size_t i = 0; i < ui_count; i++) {
+    bool tui = uis[i]->stdin_tty && uis[i]->stdout_tty;
```
any UI with `!stdin_tty` or `!stdout_tty` is considered a GUI

bfredl replied:
hmm I would change `&&` to `||` here. `cat file | nvim -` will still use the TUI and only have stdout_tty set

justinmk commented on src/nvim/ui.c:
```diff
@@ -598,6 +615,15 @@ Array ui_array(void)
+
+    // TUI fields.
+    PUT(info, "term_name", STRING_OBJ(cstr_to_string(ui->term_name)));
+    PUT(info, "term_background", STRING_OBJ(cstr_to_string(ui->term_background)));
```
Wondering if `term_background` and `stdin_fd` are too low-level to include in this UI description. Added them for completeness/consistency...

bfredl replied:
`stdin_fd` should not be included. The fd will only be used early in startup and then closed.

przepompownia commented:
@justinmk I was aware it was only draft :) Now it works as expected for me.
